### PR TITLE
Fix false positive for `no-self-argument`

### LINKED ISSUE #7300: False positive on decorators that return a `staticmethod`
### Bug description

In some usecases, a decorator is applied to a method of a python class, and it returns a staticmethod instead. Those usecases are supported by mypy with a `TYPE_CHECKING` guard: https://github.com/python/mypy/issues/6726#issuecomment-486907505.

```python
# pylint: disable=missing-docstring,invalid-name
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    returns_staticmethod = staticmethod
else:
    def returns_staticmethod(f): ...


class MyClass:
    @returns_staticmethod
    def my_staticmethod(a, b): # <======== pylint (E0213): Method should have "self" as first argument
        return a + b
```

However, testing with pylint, an error pops up indicating `self` is not the first argument of the method above. Additionally, even if we are not using the `TYPE_CHECKING` guard, pylint does not seem to  propagate `staticmethod` decorator:

```python
# pylint: disable=missing-docstring,invalid-name,too-few-public-methods
my_decorator = staticmethod

class MyClass:
    @my_decorator
    def my_decorated_func(a, b): # <======== pylint (E0213): Method should have "self" as first argument
        return a + b
```

Ideally, if pylint propagates `staticmethod` through `my_decorator`, the false positive would go away.

Thanks in advance!

### Configuration

Configuration is not relevant

### Command used

```shell
pylint main.py
```

### Pylint output

```shell
************* Module main
main.py:14:4: E0213: Method should have "self" as first argument (no-self-argument)
```

### Expected behavior

Expected that pylint propagates `staticmethod` via `my_decorator` so that the snippets below will work without warning:

```python
# pylint: disable=missing-docstring,invalid-name,too-few-public-methods
my_decorator = staticmethod

class MyClass:
    @my_decorator
    def my_decorated_func(a, b): # <======== pylint (E0213): Method should have "self" as first argument
        return a + b
```

### Pylint version

```shell
>>> pylint --version
pylint 2.12.2
astroid 2.9.0
Python 3.8.13 (default, Mar 28 2022, 11:38:47)
[GCC 7.5.0]
```

### OS / Environment

Ubuntu 20.04

### Additional dependencies

N/A

### PR DESCRIPTION
**Pull request by mbyrnepr2**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |
|    | :sparkles: New feature |
|    | :hammer: Refactoring   |
|    | :scroll: Docs          |

## Description
Fix false positive for `no-self-argument` when a staticmethod is applied to a function but uses a different name.

Closes #7300

### CODE REVIEW

File: pylint/checkers/classes/class_checker.py [hunk 1]
```diff
@@ -1926,6 +1926,9 @@ def _check_first_arg_for_type(
                 self.add_message("bad-staticmethod-argument", args=first, node=node)
                 return
             self._first_attrs[-1] = None
+        elif any(name == "builtins.staticmethod" for name in node.decoratornames()):
```
Comment (jacobtylerwalls) [inline pylint/checkers/classes/class_checker.py:1929]:
```suggestion
        elif "builtins.staticmethod" in node.decoratornames():
```

File: tests/functional/n/no/no_self_argument.py [hunk 2]
```diff
@@ -16,3 +23,18 @@ def abdc(yoo):  # [no-self-argument]
...
+        """A standard staticmethod"""
+        print("hello!")
+
+    @MYSTATICMETHOD
+    def say_goodbye():
+        """A staticmethod but using a different name"""
+        print("goodbye!")
+
+    @returns_staticmethod
+    def sum_strings(string1, string2):
```
Comment (jacobtylerwalls) [inline tests/functional/n/no/no_self_argument.py:38]:
```suggestion
    def concatenate_strings(string1, string2):
```

File: doc/whatsnew/fragments/7300.false_positive [hunk 3]
```diff
@@ -0,0 +1,3 @@
+Fix false positive for `no-self-argument` when a staticmethod is applied to a function but uses a different name.
```
Comment (jacobtylerwalls) [inline doc/whatsnew/fragments/7300.false_positive:1]:
```suggestion
Fix false positive for `no-self-argument`/`no-method-argument` when a staticmethod is applied to a function but uses a different name.
```

File: tests/functional/n/no/no_self_argument.py (hunk 2 above)
Comment (jacobtylerwalls) [reply tests/functional/n/no/no_self_argument.py:38]: Just noting for other reviewers that this test case already passed. (It's the MYSTATICMETHOD that newly passes.)

**Review by jacobtylerwalls — APPROVED** [review]
Thanks! Looks good, just some nitpicks I noticed.

File: pylint/checkers/classes/class_checker.py (hunk 1 above)
Comment (Pierre-Sassoulas) [reply pylint/checkers/classes/class_checker.py:1929]: Wow I did not see that, this is clever 😄

**Review by jacobtylerwalls — APPROVED** [review]

**Review by Pierre-Sassoulas — APPROVED** [review]

**Comment by junrushao1994** [comment]
Thank you for fixing the issue @mbyrnepr2!!

**Pull request merged**